File: Unidad 1 - Programación Funcional/ejercicio_6.hs
{- Ejercicio Nº 6: 
Definir una función que reciba una lista de listas y entregue la cantidad de elementos de la lista de mayor longitud. -}

-- Método 1: Función Básica
{-# OPTIONS_GHC -Wno-unrecognised-pragmas #-}
{-# HLINT ignore "Use foldr" #-}
cantElementos :: [[a]] -> Int
cantElementos lista = maximum ([length x | x <- lista])

-- Método 2: Función Recursiva
cantElementosRecursiva :: [[a]] -> Int
cantElementosRecursiva [] = 0
cantElementosRecursiva (x:xs) = max (length x) (cantElementosRecursiva xs)

-- Método 3: Forma Funcional
cantElementosFuncional :: [[a]] -> Int
cantElementosFuncional lista = maximum (map length lista)

-- Método 4: Función Lamda
cantElementosLambda :: [[a]] -> Int
cantElementosLambda lista = maximum (map (\x -> length x) lista)

File: Unidad 2 - Programación Lógica/Práctica/resueltos en final/ejercicio_6.pl
% Método 1: Regla Básica 
modulo(X, Y, Z):- Z is X mod Y.

% Método 2: Regla Recursiva
moduloRecursiva(X, Y, Z):- X < Y, Z is X, !.
moduloRecursiva(X, Y, Z):- X1 is X - Y, moduloRecursiva(X1, Y, Z).